notify/incidentio/incidentio.go

```diff
@@ -0,0 +1,204 @@
+		},
+	}, nil
+}
+
+// Message defines the JSON object sent to incident.io endpoints.
+type Message struct {
+	*template.Data
+
+	// The protocol version.
+	Version         string `json:"version"`
+	GroupKey        string `json:"groupKey"`
+	TruncatedAlerts uint64 `json:"truncatedAlerts"`
+}
+
+func truncateAlerts(maxAlerts uint64, alerts []*types.Alert) ([]*types.Alert, uint64) {
+	if maxAlerts != 0 && uint64(len(alerts)) > maxAlerts {
+		return alerts[:maxAlerts], uint64(len(alerts)) - maxAlerts
+	}
+
+	return alerts, 0
+}
+
+// Notify implements the Notifier interface.
+func (n *Notifier) Notify(ctx context.Context, alerts ...*types.Alert) (bool, error) {
+	alerts, numTruncated := truncateAlerts(n.conf.MaxAlerts, alerts)
+	data := notify.GetTemplateData(ctx, n.tmpl, alerts, n.logger)
+
+	groupKey, err := notify.ExtractGroupKey(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	n.logger.Debug("incident.io notification", "groupKey", groupKey)
+
+	msg := &Message{
+		Version:         "4",
+		Data:            data,
+		GroupKey:        groupKey.String(),
+		TruncatedAlerts: numTruncated,
+	}
+
+	var buf bytes.Buffer
+	if err := json.NewEncoder(&buf).Encode(msg); err != nil {
+		return false, err
+	}
+
+	var url string
+	if n.conf.URL != nil {
+		url = n.conf.URL.String()
+	} else {
+		content, err := os.ReadFile(n.conf.URLFile)
+		if err != nil {
+			return false, fmt.Errorf("read url_file: %w", err)
+		}
+		url = strings.TrimSpace(string(content))
+	}
+
+	if n.conf.Timeout > 0 {
+		postCtx, cancel := context.WithTimeoutCause(ctx, n.conf.Timeout, fmt.Errorf("configured incident.io timeout reached (%s)", n.conf.Timeout))
+		defer cancel()
+		ctx = postCtx
+	}
+
+	resp, err := notify.PostJSON(ctx, n.client, url, &buf)
+	if err != nil {
+		if ctx.Err() != nil {
+			err = fmt.Errorf("%w: %w", err, context.Cause(ctx))
+		}
+		return true, notify.RedactURL(err)
+	}
+	defer notify.Drain(resp)
+
+	shouldRetry, err := n.retrier.Check(resp.StatusCode, resp.Body)
+	if err != nil {
+		return shouldRetry, notify.NewErrorWithReason(notify.GetFailureReasonFromStatusCode(resp.StatusCode), err)
+	}
+	return shouldRetry, err
+}
+
+// errDetails extracts error details from the response for better error messages.
+func errDetails(status int, body io.Reader) string {
+	if body == nil {
+		return ""
+	}
+
+	// Try to decode the error message from JSON response
+	var errorResponse struct {
+		Message string   `json:"message"`
+		Errors  []string `json:"errors"`
+		Error   string   `json:"error"`
+	}
+
+	if err := json.NewDecoder(body).Decode(&errorResponse); err != nil {
+		return ""
+	}
+
+	// Format the error message
+	var parts []string
+	if errorResponse.Message != "" {
+		parts = append(parts, errorResponse.Message)
+	}
+	if errorResponse.Error != "" {
+		parts = append(parts, errorResponse.Error)
+	}
+	if len(errorResponse.Errors) > 0 {
+		parts = append(parts, strings.Join(errorResponse.Errors, ", "))
+	}
+
+	if len(parts) > 0 {
```

Comment:
You don't need this check, `strings.Join` is clever enough to return `""` when `len(parts)` is `0`.